include/deal.II/matrix_free/operators.h

```diff
@@ -717,6 +760,84 @@ namespace MatrixFreeOperators
     dst*= omega;
   }
 
+
+
+  //-----------------------------MassOperator----------------------------------
+
+  template <int dim, int fe_degree, int n_components, typename number>
+  MassOperator<dim, fe_degree, n_components, number>::
+  MassOperator (const MatrixFree<dim,number> &data)
+    :
+    Base<dim, number>()
+  {
+    Base<dim, number>::initialize(data);
+    compute_diagonal();
+  }
+
+
+
+  template <int dim, int fe_degree, int n_components, typename number>
+  void
+  MassOperator<dim, fe_degree, n_components, number>::
+  compute_diagonal()
+  {
+    Assert((Base<dim, number>::data != NULL), ExcNotInitialized());
+
+    LinearAlgebra::distributed::Vector<number> ones;
+    Base<dim, number>::initialize_dof_vector(Base<dim, number>::inverse_diagonal_entries);
+    Base<dim, number>::initialize_dof_vector(ones);
+    Base<dim, number>::inverse_diagonal_entries = 0.;
+    ones = 1.;
+    apply_add(Base<dim, number>::inverse_diagonal_entries, ones);
+
+    const unsigned int local_size = Base<dim, number>::inverse_diagonal_entries.local_size();
+    for (unsigned int i=0; i<local_size; ++i)
+      Base<dim, number>::inverse_diagonal_entries.local_element(i)
+        = 1./Base<dim, number>::inverse_diagonal_entries.local_element(i);
+
+    const std::vector<unsigned int> constrained_dofs
+      = Base<dim, number>::data->get_constrained_dofs();
+    for (unsigned int i=0; i<constrained_dofs.size(); ++i)
+      Base<dim, number>::inverse_diagonal_entries(constrained_dofs[i]) = 1.;
+
+    Base<dim, number>::inverse_diagonal_entries.compress(VectorOperation::insert);
```

Comment:
Since you only access locally owned elements of the vector, this operation is not necessary as there should not be anything to be done. (And we only support `LinearAlgebra::distributed::Vector` here, so it does not make sense to use something else here.)